doc/compiler/runtime/JITHooks.md

```diff
@@ -22,11 +22,49 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 
 # Background
 
-The VM has a mechansim to register callbacks for specific runtime event.
-The JIT makes use of these events to perform tasks either for functional 
-correctness, performance, or housekeeping. 
+The VM has a hook mechanism that is used to perform actions for specific
+runtime events. It also provides the ability to register callbacks for
+these events. Some hooks impact the generation of compiled code, and
+others require the JIT to perform tasks either for functional
+correctness, performance, or housekeeping. This doc outlines the various
+Hooks that are of interest for the JIT.
 
-# Hooks
+# Hooks that impact Compiled Code
+This section outlines various events or Hooks that impact the code that
+the JIT generates.
+
+**IMPORTANT**: The way the compiler tests if one of these
+events can be hooked is by calling `J9HookDisable`; if the return value
+is `0` then the event cannot be hooked, and if the return value is
+**not** `0` then the event **can** be hooked. However, it should be
+pointed out that when `J9HookDisable` returns `0`, it either successfully
+disabled the hook or the hook was already successfully disabled in some
+prior invocation. It is critical that the JIT use this API rather than
+just test whether the event could be hooked (via `J9_EVENT_IS_HOOKED`/
+`J9_EVENT_IS_RESERVED`) because of some command line option (such as
+`-Xtrace`) in order to prevent the scenario of a JVMTI agent attaching
+itself at some later point in the JVM's execution and registering
+callbacks to an event that code (generated prior to the attach) will not
+report.
```

Comment:
I'm not sure I fully follow this description.  My understanding of these sentences is rephrased as below. Is that correct? Maybe breaking down these sentences to make it bit more easier to follow? 

It is critical that the JIT uses this API rather than just tests whether the event could be hooked via `J9_EVENT_IS_HOOKED`/`J9_EVENT_IS_RESERVED`. The reason is that checking via `J9_EVENT_IS_HOOKED`/`J9_EVENT_IS_RESERVED` will not always report events registered through command line options. Such as if a JVMTI agent attaches itself at some later point in the JVM's execution and registers callbacks to an event, checking via `J9_EVENT_IS_HOOKED`/`J9_EVENT_IS_RESERVED` will not report the registered callbacks to the event.

Reply:
>The reason is that checking via J9_EVENT_IS_HOOKED/J9_EVENT_IS_RESERVED will not always report events registered through command line options.

This isn't accurate; `J9_EVENT_IS_HOOKED`/`J9_EVENT_IS_RESERVED` will always report events due to command line options. However, if NO events are registered because of command line options, then if a JVMTI agent in the future attaches to the VM, because `J9HookDisabled` wasn't called, the agent can request that the JVM report some event. However, JIT'd code would be generated assuming such an event could never occur, and now once the agent attaches itself, the JIT'd code will never report the event.

I admit, the whole thing is quite subtle and confusing; probably why the description is also kind of confusing heh. I'll try to break it up into more sentences just to make it a little easier on the reader.